source/material_model/visco_plastic.cc

```diff
@@ -191,15 +197,34 @@ namespace aspect
           // Convert friction angle from degrees to radians
           double phi = angles_internal_friction[j] * numbers::PI/180.0;
 
+          // Assing cohesions to a new variable
+          double coh = cohesions[j];
+
+          // Strain weakening
+          double strain_ii = 0;
+          if (use_strain_weakening == true)
+
+            // Calculate second strain invariant
+            strain_ii = ( (dim==3)
+                          ?
+                          ( composition[0]*composition[1] + composition[1]*composition[2] + composition[2]*composition[0] -
+                            composition[3]*composition[3] - composition[4]*composition[4] - composition[5]*composition[5]   )
+                          :
+                          ( composition[0]*composition[1] - composition[2]*composition[2] ) );
+
+          // Linear strain weakening of cohesion and internal friction angle between specified strain values
+          double strain_fraction = ( strain_ii - start_strain_weakening_intervals[j] ) /
+                                   ( start_strain_weakening_intervals[j] - end_strain_weakening_intervals[j] );
+          coh = coh + ( coh - coh * cohesion_strain_weakening_factors[j] ) * strain_fraction;
```

Comment:
Yes, the formula is a bit puzzling at first glance.   The strain fraction is always negative.  The strain_weakening_factor should be set to 1 if the user wants no strain weakening to occur. A strain_weakening_factor of 0 means that the cohesion or friction will linearly decrease to a value of 0 between the defined strain weakening interval values.  The values would reach 0 when the strain fraction is -1.  I need to describe all of this in more detail within the documentation.